### PR TITLE
[WIP] Tom clientside authorisation feedback

### PR DESCRIPTION
**Pull request by Derryrover**

### CODE REVIEW

File: src/components/AppIcon.css [hunk 1]
```diff
@@ -9,6 +9,14 @@
   float: left;
   border-radius: 2px;
 }
+.AppIcon.Disabled {
+  background-color: #737373;
+  cursor: not-allowed;
+}
+.AppIcon.Disabled img {
+  filter: grayscale(0.8);
```
[the remark this answers is not in the document]
Comment (Derryrover) [reply src/components/AppIcon.css]: "for edge: Partial support refers to supporting filter functions, but not the url function."
we are not using the url function so must be ok

File: src/components/AppIcon.css [hunk 2]
```diff
@@ -9,6 +9,14 @@
   float: left;
   border-radius: 2px;
 }
+.AppIcon.Disabled {
+  background-color: #737373;
```
[the remark this answers is not in the document]
Comment (Derryrover) [reply src/components/AppIcon.css]: Terecht commentaar.
Verschillende disabled colors aangepast in commit:
Disabled tile better greyed out color

File: src/App.js [hunk 3]
```diff
@@ -139,6 +143,27 @@ class App extends Component {
...
+      // should redirect to demo.lizard.net on prod
+      window.location = "/";
+    }
+
+    const currentHomeAppIcon = appIcons.find(icon => {
+      return window.location.href.includes(icon.linksTo.path)
+    });
+    if (currentHomeAppIcon && !doArraysHaveEqualElement(this.props.selectedOrganisation.roles, currentHomeAppIcon.requiredRoles)) {
+      const redirectMessage = this.props.intl.formatMessage({ id: "authorization.redirected_based_onrole", defaultMessage: "You do not have the rights to acces this data under the selected organisation. \nYou will be redirected" });
+      alert(redirectMessage);
```
[the remark this answers is not in the document]
Comment (Derryrover) [reply src/App.js]: Yes good point. Problem is that if user hits cancel we would need put the selected organisation back the the previous one. That makes the change a bit bigger, but still doable. This would be a good next ticket, but it has probably lower priority.
https://nelen-schuurmans.atlassian.net/browse/FRNT-513

File: src/components/OrganisationSwitcher.js [hunk 4]
```diff
@@ -70,6 +73,13 @@ class OrganisationSwitcher extends Component {
         })
       : organisations;
 
+    const currentHomeAppIcon = appIcons.find(icon => {
+      return window.location.href.includes(icon.linksTo.path)
+    });
+    const authorisationText = this.props.intl.formatMessage({ id: "authorization.organisation_not_allowed_current_page", defaultMessage: "! Organisation not authorized to visit current page !" });
```
[the remark this answers is not in the document]
Comment (Derryrover) [reply src/components/OrganisationSwitcher.js]: lets talk about this.
Is it bad enough to start troubleshooting or a new ticket for it?

File: src/home/HomeAppIconConfig.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    )
+  },
+  {
+    requiredRoles: ["admin"],
+    key: 1,
+    linksTo: {
+      external: false,
+      path: "/alarms"
+    },
+    title: <FormattedMessage id="home.alarms" defaultMessage="Alarms" />,
```
[the remark this answers is not in the document]
Comment (Derryrover) [reply src/home/HomeAppIconConfig.js]: The () are probably only needed if it starts on a newline

File: src/actions.js [hunk 6]
```diff
@@ -87,8 +87,44 @@ export function fetchOrganisations() {
...
+          };
+        })
+        const availableOrganisations = allOrganisations.filter(e => {
+          return (
+            e.roles.find(e => e === "manager") ||
+            e.roles.find(e => e === "admin") ||
+            e.roles.find(e => e === "supplier") 
+            // users with only role "user" are no longer allowed in the management page. 
+            // ||
+            // e.roles.find(e => e === "user")
```
[the remark this answers is not in the document]
Comment (Derryrover) [reply src/actions.js]: I rather have it explicitly commented-out with a comment above so the next programmer knows some history

File: src/components/OrganisationSwitcher.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Derryrover) [reply src/components/OrganisationSwitcher.js]: not bad enough to make it intoo a new ticket / action

File: src/App.js [hunk 7]
```diff
@@ -139,6 +143,27 @@ class App extends Component {
...
+
+    if ( 
+      this.props.availableOrganisations.length === 0 && 
+      this.props.isFetchingOrganisations === false &&
+      this.props.timesFetchedOrganisations > 0
+    ) {
+      const norolesMessage = this.props.intl.formatMessage({ id: "authorization.no_roles_message", defaultMessage: "Dear user, \nYou seem not to be in any organisations that can acces the management pages. \nTherefore you are redirected to the mainpage" });
+      alert(norolesMessage);
+      // should redirect to demo.lizard.net on prod
+      window.location = "/";
```
Comment (remcogerlich) [inline src/App.js]: Should it do something different on prod? I'm not convinced. Anyway either it has to do that or the comment can be removed.

File: src/App.js [hunk 8]
```diff
@@ -139,6 +143,27 @@ class App extends Component {
   };
 
   render() {
+
+    if ( 
+      this.props.availableOrganisations.length === 0 && 
+      this.props.isFetchingOrganisations === false &&
+      this.props.timesFetchedOrganisations > 0
+    ) {
+      const norolesMessage = this.props.intl.formatMessage({ id: "authorization.no_roles_message", defaultMessage: "Dear user, \nYou seem not to be in any organisations that can acces the management pages. \nTherefore you are redirected to the mainpage" });
```
Comment (remcogerlich) [inline src/App.js]: Acces -> access, and a period at the end.

File: src/App.js [hunk 9]
```diff
@@ -139,6 +143,27 @@ class App extends Component {
...
+      alert(norolesMessage);
+      // should redirect to demo.lizard.net on prod
+      window.location = "/";
+    }
+
+    const currentHomeAppIcon = appIcons.find(icon => {
+      return window.location.href.includes(icon.linksTo.path)
+    });
+    if (currentHomeAppIcon && !doArraysHaveEqualElement(this.props.selectedOrganisation.roles, currentHomeAppIcon.requiredRoles)) {
+      const redirectMessage = this.props.intl.formatMessage({ id: "authorization.redirected_based_onrole", defaultMessage: "You do not have the rights to acces this data under the selected organisation. \nYou will be redirected" });
```
Comment (remcogerlich) [inline src/App.js]: 'Access' instead of 'acces'. And isn't it better to select an organisation where the user does have the right rights? Anyway it shouldn't be possible to get into this situation.

File: src/App.js [hunk 10]
```diff
@@ -139,6 +143,27 @@ class App extends Component {
...
+      alert(redirectMessage);
+      // should redirect to demo.lizard.net/management on prod
+      this.props.history.push("/");
```
Comment (remcogerlich) [inline src/App.js]: Should go to /management/ .

File: src/actions.js [hunk 11]
```diff
@@ -87,8 +87,44 @@ export function fetchOrganisations() {
       .then(responseObj => responseObj.json())
       .then(responseData => {
         const data = responseData.results;
+        const allOrganisations = data.map(organisation => {
+          //use organisation uuid without dashes only
```
Comment (remcogerlich) [inline src/actions.js]: This is bad, IDs should be black boxes. Why was that choice made again?

File: src/actions.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (remcogerlich) [reply src/actions.js]: I disagree, commented out code just makes it harder to read. There can be a comment saying that users with only role "user" aren't allowed, but the code can be removed.

File: src/actions.js [hunk 12]
```diff
@@ -87,8 +87,44 @@ export function fetchOrganisations() {
...
+            // ||
+            // e.roles.find(e => e === "user")
+          );
+        })
+
+        dispatch({ type: RECEIVE_ORGANISATIONS, all:allOrganisations, available: availableOrganisations});
 
-        dispatch({ type: RECEIVE_ORGANISATIONS, data });
+        if (
+          !organisation ||
```
Comment (remcogerlich) [inline src/actions.js]: This variable was set in the top of this action creator (the first two lines), but here in the promise resolution the chosen organisation might have changed by now. I think it should be got from the state here.

File: src/actions.js [hunk 13]
```diff
@@ -87,8 +87,44 @@ export function fetchOrganisations() {
...
+          dispatch(
+            addNotification(
+              `Organisation "${(selectedOrganisation && selectedOrganisation.name) || 'none' }" selected`,
+              2000
+            )
+          );
+          localStorage.setItem(
+            "lizard-management-current-organisation",
+            JSON.stringify(selectedOrganisation)
+          );
```
Comment (remcogerlich) [inline src/actions.js]: Why is this here? If we do it, it should happen each time selectedOrganisation is set, and that is done in the selectOrganisation action.

Comment (remcogerlich) [reply src/actions.js]: Same holds for the notification, I think.

File: src/actions.js [hunk 14]
```diff
@@ -103,7 +139,7 @@ export function fetchOrganisations() {
 
           dispatch(
             addNotification(
-              `Organisation "${firstOrganisation.name}" selected`,
+              `Organisation "${(firstOrganisation && firstOrganisation.name) || "none"}" selected`,
```
Comment (remcogerlich) [inline src/actions.js]: Putting this comment here because only this line is changed, but: if we enter the above block, then an organisation is chosen there, and then we choose one here again. I think this block should be removed.

File: src/components/AppIcon.js [hunk 15]
```diff
@@ -2,20 +2,37 @@ import React, { Component } from "react";
...
 import { withRouter } from "react-router-dom";
+import { injectIntl } from "react-intl";
 
 
 class AppIcon extends Component {
   render() {
-    const { src, title, subTitle, handleClick } = this.props;
+    const { src, title, subTitle, handleClick, readonly, requiredRoles } = this.props;
+    const requiresRoleMessage = this.props.intl.formatMessage({ id: "authorization.requires_role", defaultMessage: "Requires role: " });
+    const requiresRolesMessage = this.props.intl.formatMessage({ id: "authorization.requires_roles", defaultMessage: "Requires one of the following roles: " });
```
Comment (remcogerlich) [inline src/components/AppIcon.js]: Doesn't react-intl have a way to deal with pluralization, besides multiple messages? Most translation libraries do.

File: src/home/App.js [hunk 16]
```diff
@@ -84,19 +50,21 @@ class App extends Component {
                 to={{ opacity: 1, x: 0 }}
                 keys={appIcons.map(item => item.key)}
               >
-                {appIcons.map((appIcon, i) => ({ x, opacity }) => (
+                {appIconsWithReadOnlyInfo.map((appIcon, i) => ({ x, opacity }) => (
```
Comment (remcogerlich) [inline src/home/App.js]: In general in map(), if we don't use the `i` parameter then it should be omitted, imo.

File: src/reducers.js [hunk 17]
```diff
@@ -49,10 +49,19 @@ function organisations(
...
+        return null;
+      }
+      let result;
+      try {
+        result = JSON.parse(localStorageResult);
+      } catch(e) {
+          return null;
+      }
+      return result;
+    }()),
```
Comment (remcogerlich) [inline src/reducers.js]: This should be an util function like `getLocalStorageJSON("lizard-management-current-organisation", null)`.

File: src/reducers.js [hunk 18]
```diff
@@ -70,38 +79,22 @@ function organisations(
...
         timesFetched: state.timesFetched + 1
       };
+    }
     case SELECT_ORGANISATION:
       // the api v3 accepts no dashes in the uuid (this is called unique_id in api v3)
       const selectedOrganisation = action.organisation;
-      selectedOrganisation.uuid = selectedOrganisation.uuid.replace(/-/g, "");
+      if (selectedOrganisation) {
+        selectedOrganisation.uuid = selectedOrganisation.uuid.replace(/-/g, "");
+      }
```
Comment (remcogerlich) [inline src/reducers.js]: Shame

File: src/translations/locales/en.json [hunk 19]
```diff
@@ -56,6 +56,11 @@
   "alarmtemplates_new.characters": "characters",
   "alarmtemplates_new.create_template": "Create template",
   "alarmtemplates_new.sms_max_char_warning": "SMS messages have a limit of 160 characters after substituting the parameter tags",
+  "authorization.redirected_based_onrole": "You do not have the rights to acces this data under the selected organisation. \nYou will be redirected",
+  "authorization.requires_role": "Requires role: ",
+  "authorization.requires_roles": "Requires one of the following roles: ",
+  "authorization.organisation_not_allowed_current_page": "! Organisation not authorized to visit current page !",
+  "authorization.no_roles_message": "Dear user, \nYou seem not to be in any organisations that can acces the management pages. \nTherefore you are redirected to the mainpage",
```
Comment (remcogerlich) [inline src/translations/locales/en.json]: Acces -> access. Mainpage -> main page

Comment (remcogerlich) [reply src/translations/locales/en.json]: Also we are very inconsistent with s/z usage, like authorization, organisation...

File: src/translations/locales/nl.json [hunk 20]
```diff
@@ -56,6 +56,11 @@
   "alarmtemplates_new.characters": "tekens",
   "alarmtemplates_new.create_template": "Sjabloon aanmaken",
   "alarmtemplates_new.sms_max_char_warning": "SMS berichten kunnen maximaal 160 tekens bevatten",
+  "authorization.redirected_based_onrole": "De geselecteerde organisatie heeft niet de rechten voor deze pagina. \nU gaat terug naar de hoofdpagina",
+  "authorization.requires_roles": "Een van de volgende rollen is nodig: ",
+  "authorization.requires_role": "vereist rol: ",
+  "authorization.organisation_not_allowed_current_page": "! Organisatie kan huidige pagina niet bezoeken !",
```
Comment (remcogerlich) [inline src/translations/locales/nl.json]: Hier kloppen zowel de eerste als de laatste zin niet. De organisatie kan dat vast wel, maar de gebruiker niet.

File: src/translations/locales/en.json [hunk 21]
```diff
@@ -56,6 +56,11 @@
   "alarmtemplates_new.characters": "characters",
   "alarmtemplates_new.create_template": "Create template",
   "alarmtemplates_new.sms_max_char_warning": "SMS messages have a limit of 160 characters after substituting the parameter tags",
+  "authorization.redirected_based_onrole": "You do not have the rights to acces this data under the selected organisation. \nYou will be redirected",
+  "authorization.requires_role": "Requires role: ",
+  "authorization.requires_roles": "Requires one of the following roles: ",
+  "authorization.organisation_not_allowed_current_page": "! Organisation not authorized to visit current page !",
```
Comment (remcogerlich) [inline src/translations/locales/en.json]: Acces -> access. Hier klopt de laatste niet (de organisation mag dat wel, maar de gebruiker niet). De eerste zin klopt wel.

File: src/App.js (hunk 8 above)
Comment (Derryrover) [reply src/App.js]: done

File: src/App.js (hunk 9 above)
Comment (Derryrover) [reply src/App.js]: Text changed.
A user can come here by url. 
Or by chaning organisation (we can discuss if this should be possible)

File: src/App.js (hunk 7 above)
Comment (Derryrover) [reply src/App.js]: How can we test this on dev?
changed it to 
// should redirect to <customer_url>.lizard.net/management/ on prod

File: src/App.js (hunk 10 above)
Comment (Derryrover) [reply src/App.js]: changed this to
// should redirect to <customer_url>.lizard.net/management/ on prod

File: src/translations/locales/en.json (hunk 19 above)
Comment (Derryrover) [reply src/translations/locales/en.json]: access changed.
Inconsistencies in organisation authorization -> can I fix this in a new pr?
This is already a problem. Not introduced in this pr

File: src/translations/locales/en.json (hunk 21 above)
Comment (Derryrover) [reply src/translations/locales/en.json]: changed

File: src/reducers.js (hunk 18 above)
Comment (Derryrover) [reply src/reducers.js]: True, but this problem is not introduced in this pr.
Last attempt to change this created a bug on production.
Can I fix this in a new pr?
https://nelen-schuurmans.atlassian.net/browse/FRNT-555

File: src/reducers.js (hunk 17 above)
Comment (Derryrover) [reply src/reducers.js]: created a ticket for this

Comment (Derryrover) [reply src/reducers.js]: done

File: src/home/App.js (hunk 16 above)
Comment (Derryrover) [reply src/home/App.js]: done

File: src/actions.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Derryrover) [reply src/actions.js]: done

File: src/actions.js (hunk 11 above)
Comment (Derryrover) [reply src/actions.js]: made ticket:
https://nelen-schuurmans.atlassian.net/browse/FRNT-555

File: src/App.js (hunk 10 above)
Comment (Derryrover) [reply src/App.js]: improved comment

File: src/App.js (hunk 9 above)
Comment (Derryrover) [reply src/App.js]: acces changed
- possible to get here by url
- or by changing organisation (should that be allowed?)

File: src/App.js (hunk 7 above)
Comment (Derryrover) [reply src/App.js]: changed comment.
difficult to test on dev

File: src/App.js (hunk 8 above)
Comment (Derryrover) [reply src/App.js]: done

File: src/actions.js (hunk 12 above)
Comment (Derryrover) [reply src/actions.js]: done moved to line 89

File: src/components/AppIcon.js (hunk 15 above)
Comment (Derryrover) [reply src/components/AppIcon.js]: changed.
I am always very hesitant to put additional logic in something like a translation library.
It adds plenty of technical complexity.

File: src/actions.js (hunk 13 above)
Comment (Derryrover) [reply src/actions.js]: Good catch !
I think this was indeed done double.
I movet it to the setOrganisation action.
Could you review once more ?

File: src/actions.js (hunk 14 above)
Comment (Derryrover) [reply src/actions.js]: true, hopefully solved in above comment
